Discontinuous spaces

TeranIvy opened the pull request:
This PR is for issue #84. I re-purposed two of the tests which were exclusively `w3` before to include `wtheta` as well, modified one test to use `w2v` discontinuous writer instead of `w3` and added `w2v` test kernel to `cma` tests on discontinuous fields. Please let me know if more tests are required.

I also updated Dynamo0.3 API documentation on fields, field vectors and function spaces, as well as a broken link to GungHo document in Introduction.

TeranIvy commented:
As mentioned in the LFRic Telco, I pushed a commit which has two fixes.

1) ` test_field_fs():` in `dynamo0p3_test.py ` now uses all function spaces as stated. It lacked `w0` before as `testkern_fs_mod.f90` had two instances of `w2`.  I changed one of them to `w0` and corrected the ` test_field_fs():` generated code. I noticed there are other instances of duplicate `w2` in `testkern_*.f90` files I modified, and I suspect those might also be the cases of forgotten `w0`. I am happy to change it if required.

2) `test_arg_discontinous():` in `dynamo0p3_test.py ` now checks for all discontinuous fields (I forgot `w2v` before).

All done until the review comments :)

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/15.1.2_builtin_and_normal_kernel_invoke.f90:
```diff
@@ -29,16 +29,17 @@
 ! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 ! POSSIBILITY OF SUCH DAMAGE.
-
+! -----------------------------------------------------------------------------
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_disc_only.f90:
```diff
@@ -31,18 +31,19 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Authors R. Ford and A. R. Porter, STFC Daresbury Lab
+! Modified I. Kavcic Met Office
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_disc_only_vector.f90:
```diff
@@ -31,18 +31,19 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Authors R. Ford and A. R. Porter, STFC Daresbury Lab
+! Modified I. Kavcic Met Office
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_w3.f90:
```diff
@@ -31,18 +31,19 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Authors R. Ford and A. R. Porter, STFC Daresbury Lab
+! Modified I. Kavcic Met Office
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/20.1.2_cma_apply_disc.f90:
```diff
@@ -2,8 +2,6 @@
 ! Copyright (c) 2017, Science and Technology Facilities Council
-! (c) The copyright relating to this work is owned jointly by the Crown,
-! Met Office and NERC 2016.
 ! However, it has been created with the help of the GungHo Consortium,
```
Please delete the three GungHo Consortium lines

TeranIvy replied:
Deleted the reference to GungHo consortium.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/20.1.2_cma_apply_disc.f90:
```diff
@@ -36,19 +34,28 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Author R. Ford and A. R. Porter, STFC Daresbury Lab
+! Modified I. Kavcic Met Office
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/4.12_multikernel_invokes_disc.f90:
```diff
@@ -31,20 +31,22 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Authors R. Ford and A. R. Porter, STFC Daresbury Lab
+! Modified I. Kavcic Met Office
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/8.2_multikernel_invokes_disc_vector.f90:
```diff
@@ -31,20 +31,22 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Authors R. W. Ford and A. R. Porter, STFC Daresbury Lab
+! Modified I. Kavcic Met Office
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/columnwise_op_app_w2v_kernel_mod.F90:
```diff
@@ -0,0 +1,135 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
Change to 2018 please

TeranIvy replied:
Changed.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/columnwise_op_app_w2v_kernel_mod.F90:
```diff
@@ -0,0 +1,135 @@
+    return
+  end function columnwise_op_app_w2v_kernel_constructor
+
+  !> @brief The subroutine which is called directly from the PSY layer and
```
I would remove the comments for brevity but I will not enforce it.

TeranIvy replied:
I left them as I think they provide some useful information.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/columnwise_op_app_w3_kernel_mod.F90:
```diff
@@ -1,21 +1,53 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_disc_only_mod.f90:
```diff
@@ -31,19 +31,24 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Authors R. Ford and A. R. Porter, STFC Daresbury Lab
+! Modified I. Kavcic Met Office
```
2017-2018 please

TeranIvy replied:
Updated the copyright.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_disc_only_vector_mod.f90:
```diff
@@ -31,19 +31,24 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Authors R. W. Ford and A. R. Porter, STFC Daresbury Lab
+! Modified I. Kavcic Met Office
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_fs_mod.f90:
```diff
@@ -0,0 +1,59 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
```
2017-2018 please

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_w3_mod.f90:
```diff
@@ -0,0 +1,82 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_write_any_w3_mod.f90:
```diff
@@ -0,0 +1,62 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_write_w2v_w1_mod.f90:
```diff
@@ -30,33 +30,32 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
-! Authors R. W. Ford and A. R. Porter, STFC Daresbury Lab
+! Author R. Ford STFC Daresbury Lab
```
This previously had A. R. Porter as an author as well as me. Could you please put him back.

TeranIvy replied:
Oops, sorry about that! He is back.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_write_w2v_w1_mod.f90:
```diff
@@ -30,33 +30,32 @@
 ! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
-! Authors R. W. Ford and A. R. Porter, STFC Daresbury Lab
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/weighted_proj_theta2_kernel_mod.F90:
```diff
@@ -1,11 +1,43 @@
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
Please delete this line as it is written further down the code.

TeranIvy replied:
Deleted.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/weighted_proj_theta2_kernel_mod.F90:
```diff
@@ -1,11 +1,43 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
2017-2018 please

TeranIvy replied:
Updated the copyright date.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -1,33 +1,41 @@
+<http://collab.metoffice.gov.uk/twiki/bin/viewfile/Static/LFRic/lfric-gungho-meto-spice/documentation/design/dynamo_formulation.pdf>`_
+and `data model
+<http://collab.metoffice.gov.uk/twiki/bin/viewfile/Static/LFRic/lfric-gungho-meto-spice/documentation/design/dynamo_datamodel.pdf>`_.
+
```
I agree and I have already asked about making the documents publicly available. For now it seems like a hassle but I will discuss it further with Steve M.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -50,15 +58,45 @@ objects and their use are discussed in the following sections.
 Please see the :ref:`algorithm-layer` section for a description of the
 ``name`` argument.
 
+Objects in Dynamo0.3 API can be categorised by their functionality
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -50,15 +58,45 @@ objects and their use are discussed in the following sections.
+Objects in Dynamo0.3 API can be categorised by their functionality
+as data types and information that specifies supported operations on
+a particular data type. The above example introduces four of five data
+types supported by Dynamo0.3 API: field, scalar, operator and column-wise
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -50,15 +58,45 @@ objects and their use are discussed in the following sections.
 +++++
 
-.. note:: To be written.
+Dynamo 0.3 API fields, identified with ``GH_FIELD`` metadata, represent
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -50,15 +58,45 @@ objects and their use are discussed in the following sections.
-.. note:: To be written.
+Dynamo 0.3 API fields, identified with ``GH_FIELD`` metadata, represent
+FEM discretisations of various dynamical core prognostic and diagnostic
+variables. In FEM variables are discretised by placing them into a
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -50,15 +58,45 @@ objects and their use are discussed in the following sections.
+Field values at points within a cell are evaluated as the sum of a set
+of basis functions multiplied by coefficients which are the data points.
+Points of evaluation are determined by a quadrature object
+(:ref:`dynamo0.3-quadrature`) and are independent of a function space
```
Changed. There are no articles in my mother tongue so I appreciate inputs from the native English speakers :)

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -50,15 +58,45 @@ objects and their use are discussed in the following sections.
 ++++++++++++
 
-.. note:: To be written.
+Depending on the function space a field lives on the field data at a
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +693,67 @@ checks (when generating the PSy layer) that any kernels which read
+
+* ``w0`` is the space of scalar functions with full continuity;
+
+* ``w1`` is the space of vector functions with full continuity;
```
The only way I could see around that was to include more information on shared dofs and where they live (vertices, edges or faces) for each function space depending on the FEM order. More text (and articles) checking for you, I am afriad :)

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -54,10 +54,9 @@
 
 # constants
-DISCONTINUOUS_FUNCTION_SPACES = ["w3"]
+DISCONTINUOUS_FUNCTION_SPACES = ["w3", "wtheta", "w2v"]
```
I agree. The tests, however, were a bit less so :)

TeranIvy commented on src/psyclone/tests/dynamo0p3_cma_test.py:
```diff
@@ -36,6 +36,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 # -----------------------------------------------------------------------------
 # Author R. Ford and A. R. Porter, STFC Daresbury Lab
+# Modified I. Kavcic, Met Office
```
Modified the copyright date and deleted the reference to GungHo consortium.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -46,9 +46,10 @@
 import pytest
```
Updated the copyright date and deleted the reference to GungHo consortium.

TeranIvy commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -3858,13 +3858,13 @@ def test_rc_discontinuous_no_depth():
 
```
Updated the copyright date.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/1.5.1_single_invoke_write_multi_fs.f90:
```diff
@@ -1,22 +1,50 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
```
Updated the copyright date.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/1.5.3_single_invoke_write_anyspace_w3.f90:
```diff
@@ -1,17 +1,44 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
```
Updated the copyright date.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/1.5_single_invoke_fs.f90:
```diff
@@ -1,22 +1,49 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
```
Updated the copyright date.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/10.9_operator_first.f90:
```diff
@@ -1,12 +1,45 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
Updated the copyright date.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/14.7_halo_annexed.f90:
```diff
@@ -1,3 +1,38 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
```
Updated the copyright date.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +693,67 @@ checks (when generating the PSy layer) that any kernels which read
+  part of ``w2``, continuous in the horizontal and discontinuous
+  in the vertical.
+
+Since Dynamo0.3 API operates on columns of data, function spaces are
```
Changed.

rupertford commented:
Hi @TeranIvy. There has been quite a big addition to master today so I'm afraid you will have to merge master back into this branch. Apologies.

TeranIvy commented on src/psyclone/tests/dynamo0p3_cma_test.py:
```diff
@@ -857,51 +858,81 @@ def test_cma_apply():
 
 
-def test_cma_apply_w3_space():
+def test_cma_apply_discontinuous_spaces():
```
Hi @rupertford! I spent all day trying to make it compile. Besides updating kernels, I also updated `argument_mod.f90` and `operator_mod.f90` to include CMA infrastructure (it was not there before).
However, it seems to me that the stripped version of LFRic infrastructure is too old and incomplete. For instance, several test files which are compiled in `dynamo0p3_test.py` use methods `fs%get_instance` and `qr%get_instance` which do not exist. Commenting out these calls does not affect compilation at all - it still works. Another test file (`10.9_operator_first.f90`) uses `function_space_collection_mod` which is not present.

Another issue is that kernel stub generation simply does not work for me (and I followed the instructions in the documentation). This is not a problem for CMA operators as I have the LFRic code to refer to, however this code is built on the updated LFRic infrastructure.

I can continue working on this, however it is going to delay the PR. Alternatively, I can open a new issue, update the LFRic infrastructure, clean up the existing compiled tests and make the modified tests compile.

TeranIvy replied:
Done after updating the branch with the latest master.

arporter commented:
@TeranIvy, Rupert's updated the stub infrastructure to support CMA as part of his redundant-computation and colouring ticket. This is now on master so you should bring your branch up-to-date and then see where we stand.

TeranIvy commented on src/psyclone/tests/dynamo0p3_cma_test.py:
```diff
@@ -857,51 +858,81 @@ def test_cma_apply():
             # The kernel only *reads* from a CMA operator and writes to a
             # field on a discontinuous space - therefore we do not need to
             # loop out into the L1 halo.
             assert "DO cell=1,mesh%get_last_edge_cell()" in code
```
Done.

rupertford commented:
@TeranIvy Sorry the addition of a compilation test took so long. You make a good point about keeping the api of our cut down infrastructure up-to-date. I think this will happen naturally as long as we have an associated compilation test.

TeranIvy commented on doc/introduction.rst:
```diff
@@ -7,8 +10,8 @@ developed to support the emerging API in the GungHo project for a
-<http://www.metoffice.gov.uk/media/pdf/g/p/MOSAC_16.10.pdf>`_ is
-designing and building the heart of the Met Office's next generation
+<https://www.metoffice.gov.uk/binaries/content/assets/mohippo/pdf/g/p/mosac_16.10.pdf>`_
+is designing and building the heart of the Met Office's next generation
```
Rewritten Introduction.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -4322,18 +4338,20 @@ def test_stencil_read_only():
     assert "a stencil must be read only" in str(excinfo.value)
 
 
-def test_w3_and_inc_error():
```
Thanks.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -6586,19 +6604,40 @@ def test_halo_for_discontinuous_2():
 
 
 def test_arg_discontinous():
```
Since child and argument indices (integer) were also changed, in the end I used dictionary with the discontinuous FS metadata as keys. From what I gathered when googling, monkeypatching is tricky when not applied to strings.

TeranIvy commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -3858,13 +3858,13 @@ def test_rc_discontinuous_no_depth():
 
 
 def test_rc_all_discontinuous_depth():
```
There were some unchanged `w3.f90` tests with quite a few `w3` writers. I renamed `disc.f90` tests and added new tests to account for different discontinuous FS. The list of modified/added tests and their dependencies is below. Hopefully the tests now have more or less even distribution of discontinuous writers. Note: for the field vectors it did not seem natural to add a vector space `w2v`.

I = dynamo0p3_test.py
II = dynamo0p3_transformations_test.py
III = dynamo0p3_cma_test.py

`w3` writers

(I ) 1.5.3_single_invoke_write_anyspace_w3.f90  <-  testkern_write_any_w3_mod.f90
(I, II) 1_single_invoke_w3.f90  <-  testkern_w3_mod.f90
(II) 1_single_invoke_w3_only_vector.f90  <-  testkern_w3_only_vector_mod.f90
(II) 8.2.1_multikernel_invokes_w3_vector.f90  <-  testkern_w3_only_vector_mod.f90
(I, II) 14.7_halo_annexed.f90  <-  testkern_w3_mod.f90
(I ) 14.7.1_halo_annexed.f90  <-  testkern_w3_mod.f90
(III) 20.1.2_cma_apply_disc.f90 (w3, w2v)  <-  columnwise_op_app_w3_kernel_mod.F90
columnwise_op_app_w3_kernel_mod.F90
testkern_w3_only_vector_mod.f90
testkern_w3_mod.f90
testkern_write_any_w3_mod.f90

`wtheta` writers

(I ) 1.5_single_invoke_fs.f90  <-  testkern_fs_mod.f90
(I, II) 1_single_invoke_wtheta.f90  <-  testkern_wtheta_mod.f90
(II) 1_single_invoke_wtheta_only_vector.f90  <-  testkern_wtheta_only_vector_mod.f90)
(II) 8.2.2_multikernel_invokes_wtheta_vector.f90  <-  testkern_theta_only_vector_mod.f90
(II) 15.1.2_builtin_and_normal_kernel_invoke.f90  <-  testkern_wtheta_mod.f90
testkern_wtheta_mod.f90
testkern_wtheta_only_vector_mod.f90 
testkern_fs_mod.f90

`w2v` writers

(I ) 1.5.1_single_invoke_write_multi_fs.f90  <-  testkern_write_w2v_w1_mod.f90
(I, II) 1_single_invoke_w2v.f90  <-  testkern_w2v_mod.f90
(II) 4.12_multikernel_invokes_w2v.f90  <-  testkern_w2v_mod.f90
(III) 20.1.2_cma_apply_disc.f90 (w3, w2v)  <-  columnwise_op_app_w2v_kernel_mod.F90
columnwise_op_app_w2v_kernel_mod.F90
testkern_w2v_mod.f90
testkern_write_w2v_w1_mod.f90

TeranIvy commented:
Hi @rupertford! I updated the tests and the examples as requested, including the transformations script in `dynamo/eg3` (I also used the opportunity to clean up the code in there a bit). 

When it comes to the LFRic examples, I lifted examples of `Wtheta` writers and put them into a new `dynamo/eg9` directory. One of the kernels (`sample_poly_adv_kernel_mod.F90`) had a commented invoke to a `Wtheta` writer as it attempted to colour a discontinuous loop. I enabled that code as `Wtheta` is now discontinuous and I am happy to report that PSyclone does not attempt to colour it. Two kernels in there have `GH_INC` access for `Wtheta` which should really be `GH_READWRITE`. As this support is being added in #25, I changed `GH_INC`  `Wtheta` to  `GH_WRITE` for now and left comments in the code. 
A side effect of this PR is that a `GH_INC` access raises an error for discontinuous FS, so I will need to finish #25 immediately after this is merged to replace `GH_INC` for instances of `Wtheta` writers in the LFRic code. This is why I put comments about introducing `GH_READWRITE` in the Dynamo0.3 documentation. Kernels in `dynamo/eg9` will serve as `GH_READWRITE` examples as well.

Finally, I changed and added quite a lot of things since the first review, so I am afraid that modifications will have to be checked again.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
+sections above are:
+
+* ``ANY_SPACE`` for when the function spaces of the modified argument(s)
+  cannot be determined;
```
Added (with some text reordering).

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -50,15 +58,45 @@ objects and their use are discussed in the following sections.
+Dynamo0.3 API fields, identified with ``GH_FIELD`` metadata, represent
+FEM discretisations of various dynamical core prognostic and diagnostic
+variables. In FEM, variables are discretised by placing them into a
+function space (:ref:`dynamo0.3-function-space`) from which they inherit
```
Sorry! Removed the "s" in the label.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -276,19 +319,19 @@ types.
-    from the function space of the modified argument. If more than one
-    argument is modified then the iteration space is taken to be the
-    largest required by any of those arguments. e.g. if a Kernel
+    from the function space of the modified argument (:ref:`dynamo0.3-function-space`).
```
Added.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -583,11 +625,11 @@ contained two calls of a kernel with arguments described by the above
+.. note:: A ``GH_FIELD`` argument that specifies ``GH_WRITE`` as its
+          access pattern must be a discontinuous function in the
+          horizontal. That means it must belong to ``w3``, ``wtheta``
+          or ``w2v`` function spaces (:ref:`dynamo0.3-function-space`).
```
Added.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
 operator values do not do so beyond the level-1 halo. If any such
 accesses are found then PSyclone aborts.
 
+.. _dynamo0.3-function-spaces:
```
Changed. Thanks for noticing!

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
+Supported Function Spaces
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As mentioned in :ref:`dynamo0.3-field` and :ref:`dynamo0.3-field-vector`
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
+onto the underlying topology and, additionally, whether the data at a
+point is a vector.
+
+Function spaces can share dofs between cells in horizontal, vertical
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
+point is a vector.
+
+Function spaces can share dofs between cells in horizontal, vertical
+or all directions. Depending on the function space and FEM order,
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
+  normal direction only. The shared dofs lie on cell faces for each
+  component.
+
+* ``w3`` is the space of scalar functions with full discontinuity.
```
Added "All dofs lie within the cell volumes and are not shared across the cell boundaries."

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
+
+Horizontally discontinuous function spaces and fields over them will not
+need colouring so PSyclone does not perform it. If such attempt is made,
+PSyclone will raise a ``Generation Error`` in **Dynamo0p3ColourTrans**
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
+need colouring so PSyclone does not perform it. If such attempt is made,
+PSyclone will raise a ``Generation Error`` in **Dynamo0p3ColourTrans**
+transformation (see :ref:`dynamo0.3-api-transformations` for more details
+on transformations). An example of field writer over a discontinuous
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
+transformation (see :ref:`dynamo0.3-api-transformations` for more details
+on transformations). An example of field writer over a discontinuous
+function space ``wtheta`` is given in ``examples/dynamo/eg9``. This
+example also demonstrates that Psyclone does not attempt to colour a
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -654,6 +700,87 @@ checks (when generating the PSy layer) that any kernels which read
+on transformations). An example of field writer over a discontinuous
+function space ``wtheta`` is given in ``examples/dynamo/eg9``. This
+example also demonstrates that Psyclone does not attempt to colour a
+loop over a discontinuous space when transformations are applied.
```
Good point. Changed.

TeranIvy commented on doc/introduction.rst:
```diff
@@ -39,12 +48,12 @@ combination thereof with no change to the algorithm or kernel layer
 PSyclone is also being extended to support an API being developed in
-the GOcean project for two finite difference ocean model benchmarks,
-one of which is based on the NEMO ocean model.
+the `GOcean project https://puma.nerc.ac.uk/trac/GOcean>`_ for two finite
```
Sorry, added it!

TeranIvy commented on examples/dynamo/README:
```diff
@@ -139,3 +140,15 @@ remove and/or change the location and depth of halo exchanges:
+    cd eg9/
+    python ../../../src/generator.py advective_inc_alg_mod.x90
+
+It also demonstrates that PSyclone does not attempt to colour loops over
```
Good point. Changed.

TeranIvy commented on examples/dynamo/eg3/colouring_and_omp.py:
```diff
@@ -1,9 +1,48 @@
-'''File containing a PSyclone transformation script for the dynamo0p3
-api to apply colouring and OpenMP generically. This can be applied via
+#!/bin/bash
```
Sorry, I must have copied it from "check_examples" together with the copyright. Deleted.

TeranIvy commented on examples/dynamo/eg9/colouring_and_omp.py:
```diff
@@ -0,0 +1,81 @@
+#!/bin/bash
```
Deleted.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/8.2.1_multikernel_invokes_w3_vector.f90:
```diff
@@ -31,20 +32,22 @@
           )
 
-end program single_invoke_w3_only_vector
+end program single_invoke_w3_only_vector
```
Added.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/8.2.2_multikernel_invokes_wtheta_vector.f90:
```diff
@@ -35,20 +31,22 @@
           )
 
-end program single_invoke_cma
+end program single_invoke_wtheta_only_vector
```
Added.

TeranIvy commented:
I hope I addressed all comments. Back to @rupertford.
